### PR TITLE
FMD-111: Database restructure

### PR DESCRIPTION
**Pull request by nsnape**
https://dluhcdigital.atlassian.net/browse/FMD-111
https://dluhcdigital.atlassian.net/browse/FMD-112
https://dluhcdigital.atlassian.net/browse/FMD-113
https://dluhcdigital.atlassian.net/browse/FMD-114

### Change description

Updates DB model definition file to include ProgrammeJunction.
Add mappings to re-route all Programme level data through this junction table.
Updates all relevant Indexes, ORM relationships, Mappings etc
Removes old FK structures including obsolete FK's to submission from Event data tables

DB / Data migration file to perform the above changes on the live database

Updates Ingest pipeline/methods to insert data into the new structure
Historical ingest pipelines updated to work with this structure.

Updates Serialisation / file download methods and queries to work with the new structure

Extensive test updates to ensure existing test logic still all works with the new structure. Including:
- Fixture changes to seed process, including source data csv mapping changes to include updated FK relations
- Serialisation tests
- Integration tests
- Extraction tests
- historical ingest/extraction tests
- DB tests (constraints, hybrid_properties, queries)

- [x] Unit tests and other appropriate tests added or updated
- [x] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
- To test app normally, pull changes, and run tests.
- To manually test (run locally), pull changes, make sure DB migrations are applied, run app. This may require a DB reset from docker-runner scripts if you have an existing DB with lots of test data in it.
- To test the data migration locally, a useful process is:
    - Switch to main branch
    - Run Flask db upgrade to previous alembic migration.
    - Add data to the DB by either/both of seed CLI or ingest endpoint
    - Download extract spreadsheet for comparison, (if required)
    - Switch back to feature branch
    - Run flask db upgrade 
    - Make sure this runs without error
    - Run app, download extract to compare (or whatever checks you want to do). The data should have migrated and for all intents and purposed appear the same as before any changes were made.
    - Optionally you could run the downgrade of the migration, and re-checkout main, and ensure it still all works the same

The above long testing process will need to be done in some detail (for QA) on a realistic test env before deploying any of these changes to prod.

### ERD (abstracted) for general DB structure after migration.

Find_ERD_proposed_V2 1

### CODE REVIEW

File: tests/db_tests/test_queries.py [hunk 1]
```diff
@@ -44,17 +45,17 @@ def outcome_data_structure_common_test(outcome_data_df):
...
 
 def test_get_download_data_no_filters(seeded_test_client, additional_test_data):
-    assert len(OutcomeData.query.all()) == 31
-    programme_with_no_projects = additional_test_data["programme_with_no_projects"]
-    programme_outcome_child_of_no_projects = additional_test_data["outcome_no_projects"]
-    assert programme_outcome_child_of_no_projects.id in [row.id for row in OutcomeData.query.all()]
+    assert len(OutcomeData.query.all()) == 30
+    # programme_with_no_projects = additional_test_data["programme_with_no_projects"]
+    # programme_outcome_child_of_no_projects = additional_test_data["outcome_no_projects"]
+    # assert programme_outcome_child_of_no_projects.id in [row.id for row in OutcomeData.query.all()]
```
Comment (cyrusdobbs) [inline tests/db_tests/test_queries.py:51]: Do we want to delete this commented out test code?

If so we can probably also delete programme_with_no_projects from additional_test_data as its not set up to be referenced by the programme junction and I can't see that it's used anywhere else

Will carry on reviewing tomorrow and test the migration locally etc 👍

File: tests/integration_tests/test_ingest_component.py [hunk 2]
```diff
@@ -209,6 +216,9 @@ def test_ingest_with_r4_file_success_with_load_re_ingest(
             "do_load": True,
         },
     )
+
+    db.session.query()
```
Comment (nsnape) [inline tests/integration_tests/test_ingest_component.py:220]: what does this do?

Comment (Joel-Pearce) [reply tests/integration_tests/test_ingest_component.py:220]: It does nothing, and I have no recollection of it entering the code. I will remove.

**Review by cyrusdobbs — APPROVED** [review]
My only suggestion is that we could look at a refactor of the load functions, some work towards this is [document here ](https://github.com/communitiesuk/funding-service-design-post-award-data-store/commit/c1d77864c70706fca5a8009824efc9a408f4c75f) but after a discussion on Slack it looks as if a refactor will be factored in after this PR. Given that this is blocking lots of merges / causing conflicts I'm happy to approve

**Comment by Joel-Pearce** [comment]
> My only suggestion is that we could look at a refactor of the load functions, some work towards this is [document here ](https://github.com/communitiesuk/funding-service-design-post-award-data-store/commit/c1d77864c70706fca5a8009824efc9a408f4c75f) but after a discussion on Slack it looks as if a refactor will be factored in after this PR. Given that this is blocking lots of merges / causing conflicts I'm happy to approve

Should we write a separate ticket for this, and do it as part of another PR you think?

File: tests/db_tests/test_queries.py [hunk 3]
```diff
@@ -44,17 +47,17 @@ def outcome_data_structure_common_test(outcome_data_df):
...
 
     # programmes with no project children should not show up even if no filters are passed
     test_query = download_data_base_query()
     test_query = test_query.with_entities(Programme.id).distinct()
 
-    test_programme_ids = [row[0] for row in test_query.all()]
-    assert programme_with_no_projects.id not in test_programme_ids
+    # test_programme_ids = [row[0] for row in test_query.all()]
+    # assert programme_with_no_projects.id not in test_programme_ids
```
Comment (gidsg) [inline tests/db_tests/test_queries.py:60]: As above can we either delete or re-instate commented out code?

File: tests/db_tests/test_queries.py [hunk 4]
```diff
@@ -82,7 +85,7 @@ def test_get_download_data_no_filters(seeded_test_client, additional_test_data):
     test_df_out = pd.read_sql(test_query_out.statement, con=db.session.connection())
 
     # programme level outcome, where the parent programme has no project children, should not show in in query.
-    assert programme_with_no_projects.programme_id not in test_df_out["programme_id"]
+    # assert programme_with_no_projects.programme_id not in test_df_out["programme_id"]
```
Comment (gidsg) [inline tests/db_tests/test_queries.py:88]: As above, can we restore or delete?

File: tests/db_tests/test_queries.py (hunk 1 above)
Comment (nsnape) [reply tests/db_tests/test_queries.py:53]: yeah, I should have removed all this. I'll remove and also remove programme_with_no_projects from the conftest data

Comment (nsnape) [reply tests/db_tests/test_queries.py:53]: done

**Review by gidsg — APPROVED** [review]

**Pull request merged**